flask_appbuilder/security/manager.py

```diff
@@ -789,186 +844,340 @@ def auth_user_db(self, username, password):
-                    self._bind_indirect_user(ldap, con)
-                    new_user = self._search_ldap(ldap, con, username)
-                    if not new_user:
-                        log.warning(LOGMSG_WAR_SEC_NOLDAP_OBJ.format(username))
+
+                # Search for `username` (if AUTH_LDAP_SEARCH is set)
+                #  - returns the `user_attributes`
+                #    needed for AUTH_USER_REGISTRATION/AUTH_ROLES_SYNC_AT_LOGIN
+                #  - we search on `username` not `bind_username`,
+                #    because AUTH_LDAP_APPEND_DOMAIN and AUTH_LDAP_USERNAME_FORMAT
+                #    would result in an invalid search filter
+                if self.auth_ldap_search:
+                    user_dn, user_attributes = self._search_ldap(ldap, con, username)
+
+                    # If search failed, go away
+                    if user_dn is None:
+                        log.info(LOGMSG_WAR_SEC_NOLDAP_OBJ.format(username))
                         return None
-                    ldap_user_info = new_user[0][1]
-                    if self.auth_user_registration and user is None:
-                        user = self.add_user(
-                            username=username,
-                            first_name=self.ldap_extract(
-                                ldap_user_info, self.auth_ldap_firstname_field, username
-                            ),
-                            last_name=self.ldap_extract(
-                                ldap_user_info, self.auth_ldap_lastname_field, username
-                            ),
-                            email=self.ldap_extract(
-                                ldap_user_info,
-                                self.auth_ldap_email_field,
-                                username + "@email.notfound",
-                            ),
-                            role=self.find_role(self.auth_user_registration_role),
-                        )
 
+            # Sync the user's roles
+            if user and user_attributes and self.auth_roles_sync_at_login:
+                user.roles = self._ldap_calculate_user_roles(user_attributes)
+                log.debug(
+                    "Calculated new roles for user='{0}' as: {1}".format(
+                        user_dn, user.roles
+                    )
+                )
+
+            # If the user is new, register them
+            if (not user) and user_attributes and self.auth_user_registration:
+                user = self.add_user(
+                    username=username,
+                    first_name=self.ldap_extract(
+                        user_attributes, self.auth_ldap_firstname_field, ""
+                    ),
+                    last_name=self.ldap_extract(
+                        user_attributes, self.auth_ldap_lastname_field, ""
+                    ),
+                    email=self.ldap_extract(
+                        user_attributes, self.auth_ldap_email_field, ""
```

Comment:
we should add `f"{username}@email.notfound" or similar, if an LDAP server does not have their email fields populated, it's impossible to register multiple users this way since it hits the email DB unique constraint

Reply:
@thesuperzapper ^^

Reply:
done

Reply:
done @dpgaspar